app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CalculatorImpl.kt

```diff
@@ -3,214 +3,121 @@ package com.simplemobiletools.calculator.helpers
+        }
     }
 
-    private fun handleRoot() {
-        mBaseValue = getDisplayedNumberAsDouble()
-        calculateResult()
+    fun handleReset() {
+        resetValues()
     }
 
-    private fun calculateResult() {
-        if (!mIsFirstOperation) {
-            updateFormula()
-        }
-
-        val operation = OperationFactory.forId(mLastOperation!!, mBaseValue, mSecondValue)
-
-        if (operation != null) {
-            updateResult(operation.getResult())
-        }
-
-        mIsFirstOperation = false
+    fun handleClear() {
+        handleReset()
     }
 
-    fun handleOperation(operation: String) {
-        if (lastKey == DIGIT && operation != ROOT)
-            handleResult()
-
-        mResetValue = true
-        lastKey = operation
-        mLastOperation = operation
-
-        if (operation == ROOT) {
-            handleRoot()
-            mResetValue = false
-        }
-
+    private fun setValue(value: String) {
+        mCallback!!.setValue(value, context)
+        displayedNumber = value
     }
 
-    fun handleClear() {
-        if (displayedNumber.equals(NAN)) {
-            handleReset()
-        } else {
-            val oldValue = displayedNumber
-            var newValue = "0"
-            val len = oldValue!!.length
-            var minLen = 1
-            if (oldValue.contains("-"))
-                minLen++
-
-            if (len > minLen) {
-                newValue = oldValue.substring(0, len - 1)
-            }
-
-            newValue = newValue.replace("\\.$".toRegex(), "")
-            newValue = formatString(newValue)
-            setValue(newValue)
-            mBaseValue = Formatter.stringToDouble(newValue)
-        }
+    private fun setFormula(value: String) {
+        mCallback!!.setFormula(value, context)
+        displayedFormula = value
     }
 
-    fun handleReset() {
-        resetValues()
+    private fun resetValues() {
+        firstNumber = 0.0
+        secondNumber = 0.0
+        decimalCounter = 0
+        decimalClicked = false
+        operator = ""
+        secondNumberSet = false
         setValue("0")
         setFormula("")
+        digits = 0
     }
 
-    fun handleEquals() {
-        if (lastKey == EQUALS)
-            calculateResult()
+    fun addDigit(i: Int) {
+        if (decimalClicked) decimalCounter--
+        if (operator != "") {
+            secondNumberSet = true
+        } else if (digits == 0) {
+            resetValues()
+        }
 
-        if (lastKey != DIGIT)
-            return
+        firstNumber = if (!decimalClicked) signumMultiply(firstNumber) * (Math.abs(firstNumber)  * 10 + i)
+        else signumMultiply(firstNumber) * (Math.abs(firstNumber) + i * Math.pow(10.0, decimalCounter.toDouble()))
+        setValue(Formatter.doubleToString(firstNumber))
 
-        mSecondValue = getDisplayedNumberAsDouble()
-        calculateResult()
-        lastKey = EQUALS
+        digits++
     }
 
     private fun decimalClicked() {
```

Comment:
This is a waste of a method. Line 31 uses it. All the other locations using decimalClicked boolean are changing it's value directly.

Reply:
fixed